### PR TITLE
Fix interactivity in auth exec

### PR DESCRIPTION
**Pull request by armandpicard**
Signed-off-by: armandpicard <armandpicard71@gmail.com>

## Motivation

Adding support for interactivity in auth exec and putting this information to the executed programs is needed; especialy on Windows (#1046 ).

## Solution

I've added simple support for `interactiveMode` by giving parent process stdin to child process if needed and give information about interactivity with the `KUBERNETES_EXEC_INFO` env variable.

### CODE REVIEW

File: kube-client/src/config/file_config.rs [hunk 1]
```diff
@@ -249,6 +249,23 @@ pub struct ExecConfig {
     /// It has been suggested in client-go via https://github.com/kubernetes/client-go/issues/1177
     #[serde(skip)]
     pub drop_env: Option<Vec<String>>,
+
+    /// Interative mode of the auth plugins
+    #[serde(rename = "interactiveMode")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub interactive_mode: Option<ExecInteractiveMode>,
```
Comment (clux) [inline kube-client/src/config/file_config.rs:256]: looks good, matches upstream!

there are a couple of extra properties related to this https://github.com/kubernetes/client-go/blob/master/tools/clientcmd/api/types.go#L257-L282 do you think they are wroth doing here while we are touching this?

**Review by clux — APPROVED** [review]
LGTM. Thanks a lot! Leaving it open for comments for a bit.

File: kube-client/src/client/auth/mod.rs [hunk 2]
```diff
@@ -64,6 +64,10 @@ pub enum Error {
     #[error("failed to parse auth exec output: {0}")]
     AuthExecParse(#[source] serde_json::Error),
 
+    /// Fail to serialize input
+    #[error("fail to serialize input: {0}")]
```
Comment (kazk) [inline kube-client/src/client/auth/mod.rs:68]:
```suggestion
    #[error("failed to serialize input: {0}")]
```

Comment (clux) [reply kube-client/src/client/auth/mod.rs]: Applied and synced the branch here.

File: kube-client/src/client/auth/mod.rs [hunk 3]
```diff
@@ -500,6 +508,25 @@ fn auth_exec(auth: &ExecConfig) -> Result<ExecCredential, Error> {
         cmd.envs(envs);
     }
 
+    let interactive = auth.interactive_mode != Some(ExecInteractiveMode::Never);
+    if interactive {
+        cmd.stdin(std::process::Stdio::inherit());
+    } else {
+        cmd.stdin(std::process::Stdio::piped());
+    }
```
Comment (clux) [inline kube-client/src/client/auth/mod.rs:516]: Double checked this with [std::process::Command::stdin](https://doc.rust-lang.org/std/process/struct.Command.html#method.stdin) docs and this should indeed be a non-breaking addition.

File: kube-client/src/config/file_config.rs (hunk 1 above)
Comment (clux) [reply kube-client/src/config/file_config.rs:262]: resolving, this stands on its own

File: kube-client/src/config/file_config.rs [hunk 4]
```diff
@@ -249,6 +249,23 @@ pub struct ExecConfig {
     /// It has been suggested in client-go via https://github.com/kubernetes/client-go/issues/1177
     #[serde(skip)]
     pub drop_env: Option<Vec<String>>,
+
+    /// Interative mode of the auth plugins
```
[the remark this answers is not in the document]
Comment (clux) [reply kube-client/src/config/file_config.rs:259]: upstream comment is not that great either way, resolving this.

**Comment by clux** [comment]
Merging this, it's a clean non-breaking change judging by the Command docs. Sorry for the delay, and thanks for the contribution!

**Pull request merged**